psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1498 @@
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/oeprop.h"
+#include "psi4/libmints/twobody.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libpsi4util/PsiOutStream.h"
+#include "psi4/libpsi4util/process.h"
+#include "psi4/libpsio/psio.h"
+#include "psi4/libpsio/psio.hpp"
+#include "psi4/libqt/qt.h"
+#include "psi4/libciomr/libciomr.h"
+
+#include <algorithm>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+
+namespace psi {
+namespace dlpno {
+
+DLPNOMP2::DLPNOMP2(SharedWavefunction ref_wfn, Options& options) : Wavefunction(options) {
+    shallow_copy(ref_wfn);
+    reference_wavefunction_ = ref_wfn;
+
+    common_init();
+}
+DLPNOMP2::~DLPNOMP2() {}
+void DLPNOMP2::common_init() {
+    print_ = options_.get_int("PRINT");
+    debug_ = options_.get_int("DEBUG");
+
+    name_ = "DLPNO-MP2";
+    module_ = "dlpno";
+
+    variables_["SCF TOTAL ENERGY"] = reference_wavefunction_->energy();
+
+    ribasis_ = get_basisset("DF_BASIS_MP2");
+}
+
+/* Utility function for easily performing making C_DGESV calls
+ *
+ * C_DGESV solves AX=B for X, given symmetric NxN matrix A and NxM matrix B
+ * B is expected in fortran layout, which complicates the call when (M > 1)
+ * The workaround used here is switching layouts of B before and after the call
+ */
+void C_DGESV_wrapper(SharedMatrix A, SharedMatrix B) {
+
+    int N = B->rowspi(0);
+    int M = B->colspi(0);
+    if(N == 0 || M == 0) return;
+
+    // create a copy of B in fortran ordering
+    double* B_fortran = new double[N * M];
+    for(int n = 0; n < N; n++) {
+        for(int m = 0; m < M; m++) {
+            B_fortran[m * N + n] = B->get(n, m);
+        }
+    }
+
+    // make the C_DGESV call, solving AX=B for X
+    int *ipiv = init_int_array(N);
+    int errcode = C_DGESV(N, M, A->pointer()[0], N, ipiv, B_fortran, N);
+    free(ipiv);
+
+    // copy the fortran-ordered X into the original matrix, reverting to C-ordering
+    for(int n = 0; n < N; n++) {
+        for(int m = 0; m < M; m++) {
+            B->set(n, m, B_fortran[m * N + n]);
+        }
+    }
+
+}
+
+/*
+ * In order to use DIIS to accelerate convergence of the MP2 amplitudes,
+ * we need to "flatten" the sparse data structure (list of Matrix objects)
+ * into a single Matrix
+ */
+SharedMatrix flatten_mats(const std::vector<SharedMatrix> &mat_list) {
+
+    size_t total_size = 0;
+    for(SharedMatrix mat : mat_list) {
+        total_size += (mat->colspi(0) * mat->rowspi(0));
+    }
+    //outfile->Printf("total_size: %zu\n",total_size);
+
+    SharedMatrix flat = std::make_shared<Matrix>("flattened matrix list", 1, total_size);
+    //outfile->Printf("made it!\n");
+    size_t ind = 0;
+
+    for(SharedMatrix mat : mat_list) {
+        //outfile->Printf("%d %d\n", mat->rowspi(0), mat->colspi(0));
+        for(size_t r = 0; r < mat->rowspi(0); r++) {
+            for(size_t c = 0; c < mat->colspi(0); c++, ind++) {
+                flat->set(0, ind, mat->get(r,c));
```

Comment:
You can directly access the memory pointer to the `flat` and `mat` data and replace these two inner for loops with a `memcpy` call.  That would be more efficient than these for loops.

Reply:
Good point, thanks!